### ISSUE TITLE
0.10.20 breaks saves on populated models

### ISSUE THREAD

**Issue by particlebanana** [balderdashy/waterline#937]
I know when we merged #924 we were removing the ability to call `.save` on individual populated collection items but didn't think to check populated model associations. So in the classic pet user example:

```javascript
User.findOne({ id: 123 })
.populate('pet')
.exec(function(err, user) {
  var pet = user.pet;
  pet.name = 'fido';
  pet.save();
});
```

If it wasn't documented anywhere that's fine but we should check for it so it doesn't throw.

edit: crashing here https://github.com/balderdashy/waterline/blob/master/lib/waterline/model/lib/defaultMethods/save.js#L47

**Comment by devinivy** [comment]
Ah!  Relatedly: https://github.com/balderdashy/waterline/pull/829#issuecomment-74443706

**Comment by particlebanana** [comment]
Trying to write a test for it and I can't seem to re-create it. Stack trace if anyone figures it out in the meantime.

```javascript
/app/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:47 
var modelKeys = Object.keys(proto.associationsCache); 
                       ^ 
TypeError: Object.keys called on non-object 
    at Function.keys (native) 
    at Array.async.auto.compareModelValues [as 0] (/app/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:47:30) 
    at /app/node_modules/sails/node_modules/async/lib/async.js:487:38 
    at Array.forEach (native) 
    at _each (/app/node_modules/sails/node_modules/async/lib/async.js:46:24) 
    at Object.async.auto (/app/node_modules/sails/node_modules/async/lib/async.js:458:9) 
    at new module.exports (/app/node_modules/sails/node_modules/waterline/lib/waterline/model/lib/defaultMethods/save.js:42:9) 
    at prototypeFns.save (/app/node_modules/sails/node_modules/waterline/lib/waterline/model/index.js:34:14) 
```

**Comment by blah238** [comment]
Sorry if this got broken. Perhaps what we need is to differentiate read-only finds (for better performance/reduced memory overhead) and update-able finds with association caching or whatever zany business is going on in there. Either way, populating still takes up way too much memory.

**Comment by particlebanana** [comment]
I'm down for removing the `.save` instance method if it adds too much overhead but we can't break stuff in `0.10` patches so that's what I'm worried about. I'm working on a test in wl adapter tests. Looks like we weren't testing `.save` at all in there.

**Comment by tjwebb** [comment]
fyi, `_.keys(null)` does not fail.

Yea ideally we don't want to break stuff in patches, but there's no way to know right now. We're evolving the waterline-adapter-tests very rapidly and we're improving coverage every day. Whatever conditions produce this issue will be added to waterline-adapter-tests

**Comment by dmarcelino** [comment]
>  Looks like we weren't testing .save at all in there.

@particlebanana, I [searched for save](https://github.com/balderdashy/waterline-adapter-tests/search?utf8=✓&q=save) in the wl adapter tests and I can see we only test it in the context of `.add()` and `.remove()`.

Meanwhile I've a added a test for this and was able to reproduce the error [Object.keys called on non-object](https://travis-ci.org/balderdashy/waterline-adapter-tests/jobs/57631980#L247).
Check PR  balderdashy/waterline-adapter-tests#46.

**Comment by dmarcelino** [comment]
A potential path to fix this issue is to partially revert PR #829, I've committed aa29dcc for you to take a look. This passes the adapter tests in PR balderdashy/waterline-adapter-tests#46. The only caveat of this is that we increase the memory footprint that PR #829 aimed to fix.

**Comment by particlebanana** [comment]
I say let's revert it for now simply because if there are people using it, even undocumented, it's been stable for 19 releases. Let's add it back in 0.11 and see if we can look at ways to further decrease the populated footprint.

:+1:  on the adapter tests @dmarcelino those things are getting nice!

**Comment by dmarcelino** [comment]
Agreed, issued PR #938.

Thanks @particlebanana, recently there have been some very good contributions towards it. Talking of which, I would really like to get your attention on balderdashy/waterline-adapter-tests#41 which allows to fix the breaking tests on sails-memory and sails-disk ;)

**Comment by dmarcelino** [comment]
@devinivy, @tjwebb, @particlebanana, on the topic of bug squashing and stability, we should seriously look into the following for `0.10.21`:
* this one, obviously: #938 / balderdashy/waterline-adapter-tests#46
* Use correct foreign collection during nested update creation/link step: #934 / balderdashy/waterline#930
* Fix crash for nested model create within update query: #930
* Add Interface SQL (fixes sails-memory and sails-disks false positives): balderdashy/waterline-adapter-tests#41

In my view these are the current top priorities.

**Comment by particlebanana** [comment]
* #938 / https://github.com/balderdashy/waterline-adapter-tests/pull/46 - looks great
*  https://github.com/balderdashy/waterline-adapter-tests/pull/41 - looks great

As for #934 #930 I'll have to read up on them some. It sounds like it's just an issue of using the attribute name instead of the referenced model right?

**Comment by particlebanana** [comment]
I don't see anything bad about #934 as a standalone PR and if https://github.com/balderdashy/waterline-adapter-tests/pull/44 passes all the tests afterwards that is golden. Thanks for that work on refactoring the tests @devinivy!

Still going through #930

**Comment by devinivy** [comment]
@particlebanana the most important thing about #934 right now is that there's a claim that nested updates duplicates the `createBelongsTo` routine (https://github.com/balderdashy/waterline/pull/930#issuecomment-90324885).  To stop that duplication, #930 seems to be a reasonable solution.

**Comment by dmarcelino** [comment]
Merged both #938 and balderdashy/waterline-adapter-tests#46, so this issue can be considered fixed.

@blah238, @nbiton, given I partially reverted #924 we've lost some of the memory/speed improvements but I'm sure everyone will welcome another effort! Besides, now we have more tests to cover potential issues! :smiley:

**Comment by particlebanana** [comment]
@devinivy @dmarcelino #934 and #930 should be good. They work together and should be fine if they pass all the tests.
